### ISSUE TITLE
Wav2vec2.0 Inference with Transformer LM gives worse results than with KenLM or viterbi

### ISSUE THREAD

**Issue by thanhkm** [pytorch/fairseq#3362]
#### What is your question?
I do inference with Transformer LM but it results is worse than the one with KenLM or Viterbi.

**WER on test-other:
Viterbi: 8.42 KenLM: 6.90 transformerLM: 10.48**

Following the wav2vec2 paper, the results with Transformer LM should be better than the others. Am I missing something ?

#### Code

python3 examples/speech_recognition/infer.py /path/to/librispeech/ --task audio_pretraining --nbest 1 --path /path/to/wav2vec_smal
l_960h.pt --gen-subset test-other --w2l-decoder fairseqlm --lm-model /path/to/lm_librispeech_word_transformer.pt --lm-weight 2 --word-score -1 --sil-weight 0 --criterion ctc --labels ltr --max-tokens 4000000 --post-process letter --lexicon /path/to/dict.lst --beam=8

#### What have you tried?

- wav2vec model: wav2vec2 base finetuned on 960hours https://dl.fbaipublicfiles.com/fairseq/wav2vec/wav2vec_small_960h.pt
- transformer models: taken from wav2letter 2019 sota as sugesstion in the docs https://dl.fbaipublicfiles.com/wav2letter/sota/2019/lm/lm_librispeech_word_transformer.pt
- dictionary: taken from wav2letter 2019 sota https://dl.fbaipublicfiles.com/wav2letter/sota/2019/lm/lm_librispeech_word_transformer.dict

#### What's your environment?

 - fairseq Version : 1.0.0a0+d031611
 - flashlight Version: build from master
 - PyTorch Version : 1.8.0+cu111
 - OS : Linux
 - Python version: 3.6.12
 - CUDA/cuDNN version: 11.2
 - GPU models and configuration: RTX 3080
 - Any other relevant information:

**Comment by Guruprasad68** [comment]
Check the wav2vec2 paper(ablation section) for value of correct LM weight, word insertion penalty and beam. Beam of 8 is too small, i guess they had used 500 for Transformer LM. Matching those gives the correct results.

**Comment by edosyhptra** [comment]
I did inference with Viterbi and got result in 2.549

`2021-06-28 02:59:15 | INFO | valid | {"epoch": 196, "valid_loss": "10.59", "valid_ntokens": "155.374", "valid_nsentences": "5.84173", "valid_nll_loss": "0.398", "valid_uer": "1.743", "valid_wer": "2.549", "valid_raw_wer": "2.549", "valid_wps": "1967.1", "valid_wpb": "155.4", "valid_bsz": "5.8", "valid_num_updates": "85752", "valid_best_wer": "2.495"}`

My question is, is that 2.549 means 2% of WER or 20%? How do we measure that value?

**Comment by RuABraun** [comment]
that's 2.5% WER

**Comment by thanhkm** [comment]
> Check the wav2vec2 paper(ablation section) for value of correct LM weight, word insertion penalty and beam. Beam of 8 is too small, i guess they had used 500 for Transformer LM. Matching those gives the correct results.

I means with the help of an extra LM, it should improve performance. This is the case for KenLM but not for TransformerLM. In addition, TransformerLM is more powerful than KenLM. There should be something wrong ... 





> `2021-06-28 02:59:15 | INFO | valid | {"epoch": 196, "valid_loss": "10.59", "valid_ntokens": "155.374", "valid_nsentences": "5.84173", "valid_nll_loss": "0.398", "valid_uer": "1.743", "valid_wer": "2.549", "valid_raw_wer": "2.549", "valid_wps": "1967.1", "valid_wpb": "155.4", "valid_bsz": "5.8", "valid_num_updates": "85752", "valid_best_wer": "2.495"}`
> 
> My question is, is that 2.549 means 2% of WER or 20%? How do we measure that value?

It's 2.549% WER.
